src/opustrainer/modifiers/typos.py

```diff
@@ -0,0 +1,164 @@
+    """Special version of typo's random_space and skipped_space that also
+    updates alignment info.
+    
+    action: add | remove
+      whether to add or remove a random space
+
+    strval: str
+      input text
+
+    alignments: str
+      string of space split m-n pairs
+
+    column: int
+      either 0 or 1, decides which side of the alignment pairs is updated
+    """
+    # all the locations where there are non-space characters.
+    locations = [m.start() for m in re.finditer(r'\S', strval)]
+    
+    if len(locations) == 0:
+        return strval, alignments
+
+    # Select character after which to add a space
+    char_index = locations[random.randint(0, len(locations) - 1)]
+
+    # Figure out which word that character falls in
+    word_index = sum(1 for _ in re.finditer(r'\s+', strval[:char_index]))
+
+    # Insert space
+    if action == 'add':
+        strval = strval[:char_index] + ' ' + strval[char_index:]
+    else:
+        strval = strval[:char_index-1] + strval[char_index:]
+
+    # Fix up alignments
+    fixed_alignments = []
+    for alignment in alignments.split(' '):
+        # Splits the a-b pairs into tuples
+        mapping = [int(index) for index in alignment.split('-', maxsplit=1)]
+
+        # Alignments before the introduced space stay as-is. Intentionally, if
+        # the mapping is about word_index itself, we apply both to duplicate
+        # the mapping.
+        if mapping[column] <= word_index:
+            fixed_alignments.append(f'{mapping[0]}-{mapping[1]}')
+        
+        # Alignments after the space are shifted by 1
+        if action == 'add' and mapping[column] >= word_index \
+           or action == 'remove' and mapping[column] > word_index:
+            if action == 'add':
+                mapping[column] += 1
+            else:
+                mapping[column] -= 1
+            fixed_alignments.append(f'{mapping[0]}-{mapping[1]}')
+
+    return strval, ' '.join(fixed_alignments)
+
+
+class TypoModifier(Modifier):
+    modifiers = {
+        'char_swap':     0.1,
+        'missing_char':  0.1,
+        'extra_char':    0.1,
+        'nearby_char':   0.1,
+        'similar_char':  0.1,
+        'skipped_space': 0.1,
+        'random_space':  0.1,
+        'repeated_char': 0.1,
+        'unichar':       0.1,
+    }
+
+    column: int
+
+    probabilities: Dict[str,float]
+
+    def __init__(self, probability:float, column:Literal['src', 'trg']='src', **probabilities:float):
+        """
+        Apply typo modifiers to the input. If no specific typo modifiers are
+        mentioned, it will default to applying them all with a 0.1 probability
+        each. If modifiers are mentioned in the configuration, only the
+        modifiers mentioned will be used. All probabilities have to be in the
+        0.0 .. 1.0 range.
+
+        args:
+            probability: float
+                probability a line will be modified
+
+            column: 'src' | 'trg'
```

Comment:
Just a thought... Why would we ever want to corrupt the target?

Reply:
Very good question.